homeassistant/components/neato/__init__.py

```diff
@@ -103,7 +104,7 @@ async def async_setup_entry(hass, entry):
         await hass.async_add_executor_job(hub.update_robots)
     except NeatoRobotException:
         _LOGGER.debug("Failed to connect to Neato API")
-        return False
+        raise ConfigEntryNotReady
```

Comment:
You should make sure that you don't make changes to `hass.data` until after you can raise ConfigEntryNotReady. Otherwise it will keep the reference around.

Reply:
So the first check that is done up above is where we check the login credentials.  If they are invalid we still `return False` there correctly.  Here I am raising config entry not ready because the second endpoint from neato fails to connect and if we retry with the same data it will work.

Reply:
Paulus means this:

```py
    hub = NeatoHub(hass, entry.data, Account)

    await hass.async_add_executor_job(hub.login)
    if not hub.logged_in:
        _LOGGER.debug("Failed to login to Neato API")
        return False

    try:
        await hass.async_add_executor_job(hub.update_robots)
    except NeatoRobotException:
        _LOGGER.debug("Failed to connect to Neato API")
        raise ConfigEntryNotReady

    hass.data[NEATO_LOGIN] = hub
```

Reply:
Thank you so much @frenck !  This is now corrected.